returnn/util/file_cache.py

```diff
@@ -203,45 +203,79 @@ def cleanup(self, *, need_at_least_free_space_size: int = 0):
         for mtime, neg_size, fn in all_files:
             size = -neg_size
             delete_reason = None
-            if cur_time - mtime > self._cleanup_files_always_older_than_days * 60 * 60 * 24:
-                delete_reason = f"File is {(cur_time - mtime) / 60 / 60 / 24:.1f} days old"
-            else:
-                reached_more_recent_files = True
-            if not delete_reason and need_at_least_free_space_size > cur_expected_free:
-                # Still must delete some files.
-                if cur_time - mtime > cur_used_time_threshold:
-                    delete_reason = f"Still need more space, file is {(cur_time - mtime) / 60 / 60:.1f} hours old"
-                else:
-                    raise Exception(
-                        f"We cannot free enough space on {self.cache_directory}.\n"
-                        f"Needed: {human_bytes_size(need_at_least_free_space_size)},\n"
-                        f"currently available: {human_bytes_size(cur_expected_free)},\n"
-                        f"oldest file is still too recent: {fn}.\n"
-                        f"{report_size_str}"
-                    )
-            if not delete_reason and want_free_space_size > cur_expected_free:
-                if cur_time - mtime > self._cleanup_files_wanted_older_than_days * 60 * 60 * 24:
-                    delete_reason = f"Still want more space, file is {(cur_time - mtime) / 60 / 60:.1f} hours old"
-                else:
-                    # All further files are even more recent, so we would neither cleanup them,
-                    # so we can also just stop now.
-                    break
 
-            if delete_reason:
-                cur_expected_free += size
-                print(
-                    f"FileCache: Delete file {fn}, size {human_bytes_size(size)}. {delete_reason}."
-                    f" After deletion, have {human_bytes_size(cur_expected_free)} free space."
-                )
+            lock_dir, lock_file = self._get_lock_filename(fn)
+            with LockFile(directory=lock_dir, name=lock_file, lock_timeout=self._lock_timeout):
```

Comment:
I realize, we would not want the code like this in `cleanup`. This would potentially hang for all files which are currently copied by other procs, right? The copying could potentially take some time. The `cleanup` code should just skip over those files where there is an active lock.

To do this in a clean way, maybe our `LockFile` API needs to be extended. E.g. there could be some `LockFile.try_lock_ctx` or so.

Reply:
I think `try_lock_ctx` is not going to work as we wouldn't want the block contents to be executed when the lock isn't held, but a normal try-finally and an if will do.

Reply:
> I think `try_lock_ctx` is not going to work as we wouldn't want the block contents to be executed

I was thinking about sth like:
```python
with LockFile(...).try_lock_ctx() as lock_ctx:
    if not lock_ctx.we_hold_the_lock():
        continue
    ...
```
But your code is also fine now.